### DISCUSSION TITLE
Questions of "message persistence" on v5.0 document

### DISCUSSION THREAD

**Discussion by Nextdrive-IanChiu** [emqx/emqx#8623] [Q&A] [answered]
Hi,

I have a question about MQTT persistent session and offline message when QoS > 0.
I found there are saying EQMX open source doesn't support message persistence in [v5.0 document](https://www.emqx.io/docs/en/v5.0/design/design.html#message-persistence).

However, from the tests I did and these issues ([issue1](https://github.com/emqx/emqx/issues/3817), [issue2](https://github.com/emqx/emqx/issues/3577), and [issue3](https://github.com/emqx/emqx/issues/3529)), it seems like we could use the persistent session and get offline messages in open source version.

Do I miss something or do I have a wrong understanding of the v5.0 document?

**Comment by ieQu1** [comment, marked as answer]
Hi,
Thanks for pointing this out, the document should be updated indeed. The current implementation of message persistence in 5.0 is experimental though, and we at EMQ are not quite happy with its performance. In 5.1 we're planning to roll out an optimized implementation.

**Reply by Nextdrive-IanChiu** [reply]
Thanks for your help. Looking forward to the new implementation in version 5.1!

**Comment by xzycn** [comment]
@IanChiu628 
How can you get "persistent session and offline message"?
I‘m runing version 5.0.4.
My steps as follows:
1、Running EMQX with Docker(mount local path： -v /opt/emqx/data:/opt/emqx/data)
2、Running MQTTX
     1）Create a connection to the EMQX
     2）Publish a mssage to topic /msg/channel/1 with QoS 1
     3）Restart Docker container emqx
     4）Create a subscription with topic /msg/channel/1 above the previous connection

As a result: the subscription  cannot recieve the message.
Did I miss something? How can I debug this?

Thank you in advance :)

**Reply by Nextdrive-IanChiu** [reply]
I think you shouldn't restart emqx in step `3 )`. From my understanding, "persistent session and offline message" are stored on MQTT broker. So the mechanism won't work at all if the MQTT broker is restarted.

Maybe you can find [this blog](https://www.emqx.com/en/blog/mqtt-session) for reference. 😄

**Reply by xzycn** [reply]
Thank you for your blog. Actually I care about the message. If the messages would have gone away after restarting emqx,it shouldn't call "message persistence" in my opinion :(
I want to persist the message in my scenario,do you have a suggestion for it? It seems that "[Data storage](https://www.emqx.io/docs/en/v4.4/backend/backend.html)" is the way,but it do not support most bakends like mysql、kafka in open-source version :(

**Reply by Nextdrive-IanChiu** [reply]
What you say is out of the MQTT protocol scope. Maybe we can only rely on MQTT broker solutions and the feature is only available on the enterprise version of EMQX.

**Reply by ieQu1** [reply]
Hello @zxcyn,

Persistent layer in 5.0 is disabled by default. 
In order to enable it, you need to run EMQX container with the following parameters:

```
    EMQX_PERSISTENT_SESSION_STORE__ENABLED=true \
    EMQX_PERSISTENT_SESSION_STORE__BACKEND__SESSION__RAM_CACHE=false \
    EMQX_PERSISTENT_SESSION_STORE__BACKEND__SESSION_MESSAGES__RAM_CACHE=false \
    EMQX_PERSISTENT_SESSION_STORE__BACKEND__MESSAGES__RAM_CACHE=false \
```
